fvm/evm/precompiles/arch.go

```diff
@@ -0,0 +1,52 @@
+package precompiles
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	gethCommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/onflow/flow-go/fvm/evm/types"
+)
+
+var (
+	FlowBlockHeightFuncSig = ComputeFunctionSignature("flowBlockHeight", nil)
+	// TODO update me with a higher value if needed
+	FlowBlockHeightFixedGas = uint64(1)
+)
+
+// ArchContract return a procompile for the Cadence Arch contract
+func ArchContract(
+	address types.Address,
+	heightProvider func() (uint64, error),
+) types.Precompile {
+	return MultiFunctionPrecompileContract(
+		address,
+		[]Function{&flowBlockHeightFunction{heightProvider}},
+	)
+}
+
+type flowBlockHeightFunction struct {
+	flowBlockHeightLookUp func() (uint64, error)
+}
+
+func (c *flowBlockHeightFunction) FunctionSignature() FunctionSignature {
+	return FlowBlockHeightFuncSig
+}
+
+func (c *flowBlockHeightFunction) ComputeGas(input []byte) uint64 {
+	return FlowBlockHeightFixedGas
+}
+
+func (c *flowBlockHeightFunction) Run(input []byte) ([]byte, error) {
+	if len(input) > 0 {
+		return nil, fmt.Errorf("unexpected input is provided")
+	}
+	bh, err := c.flowBlockHeightLookUp()
+	if err != nil {
+		return nil, err
+	}
+	encoded := make([]byte, 8)
+	binary.BigEndian.PutUint64(encoded, bh)
+	return gethCommon.LeftPadBytes(encoded, 32), nil
```

Comment:
Why 32? because all values are actually 256 bit?

Reply:
I found this convention in the Geth precompile codes to always return 32 bytes, even for booleans they return a 32 byte value, I guess this has to do with how get uses 32 bytes for returning anything and later conversion.

Reply:
I believe the reason for it is EVM natively works with 256bit words under the hood and any casting etc costs gas, so they optimize it by returning a left padded value to minimize the gas on the consumer side

Reply:
added a comment about it in the code.